### ISSUE TITLE
Error with iOS Hosted Content

### ISSUE THREAD

**Issue by G3n3ralTao** [j3k0/cordova-plugin-purchase#281]
I got some trouble with non-consumable purchase with content hosted on Apple servers.
The InApp are created in my account and I uploaded the content via ApplicationLoader and everything is alright on the iTunes page. I created this small js function to test the InApp:

;

store.verbosity = store.DEBUG;

        store.register({
            id:    'com.test.testApp.testWithData', 
            alias: 'hosted content download',
            type:   store.NON_CONSUMABLE
        });

        store.refresh();

        /*
         * iOS Apple-hosted content
         */

        // When purchase of the downloadable content is approved,
        // show some logs.
        store.when("hosted content download").approved(function (product) {
            console.log("You've purchased the content for " + product.id + " - it will now download to your device!");
            console.log(product);
        });

        // Show progress during hosted content download
        store.when("hosted content download").downloading(function(product) {
            console.log(product.id + " downloading!");

        });

        // When hosted content download is complete, finish the transaction
        store.when("hosted content download").downloaded(function(product) {
            console.log(product.id + " downloaded!");
            product.finish();
        });

        // If the product content is downloading or downloaded, display the downloaded content
        store.when("hosted content download").updated(function (product) {
            console.log(product.id + " updated!");
            console.log(product);
        });

        store.refresh();

        store.order("hosted content download");

I get an error when I try to bind callback on the functions downloading and downloaded:
"TypeError: undefined is not a function (store.when("hosted content download").downloaded(function(product) {console.log(product.id + " downloaded!");product.finish();}))"

The update callback is never called and I never get any notification that something is downloaded.

I based myself on the demo at this address: "https://github.com/dpa99c/cordova-plugin-purchase-demo/blob/master/www/js/index.js". 

Anyone know what is going on?

**Comment by dpa99c** [comment]
If `store.downloaded` is not defined, I'd suspect that there's an issue with the plugin JS being included into the app. As you can see on [line 274 of store-ios.js](https://github.com/j3k0/cordova-plugin-purchase/blob/master/www/store-ios.js#L274), the `downloaded` promise is added.

You can try using the [remote debugging functionality of Safari](http://phonegap-tips.com/articles/debugging-ios-phonegap-apps-with-safaris-web-inspector.html) to step through the JS line-by-line and confirm what's causing the error.

**Comment by G3n3ralTao** [comment]
I opened the file in the plugin folder of my project and the promise are not there. This is really strange since when I look at the plugin version I am up to date. I will dig a little further on why the version is wrong and test it directly with the script from github. Thank for the pointer.

**Comment by G3n3ralTao** [comment]
Update

I removed/added the plugin with "cordova plugin add cc.fovea.cordova.purchase" and I have two new problems:

1) The store-ios.js file in my plugin folder still doesn't have the 2 missing promises. That means that there is something wrong with the updating system at some point or I am using the wrong adress for some reasons

2) When I use the "cordova build ios" command, the build always fail and I get error exit code 2. My Cordova version is 5.1.1. Is it a factor?

I also tried to just replace the file directly with the one on github and it's not working either. The app doesn't start.

**Comment by G3n3ralTao** [comment]
Update 2

Just to be sure, I created a new cordova project and I added directly the InApp plugin via "cordova plugin add cc.fovea.cordova.purchase" and the file I get doesn't have the 2 promises. Something is wrong somewhere. Is there a way I can make a manual installation of the plugin?

**Comment by G3n3ralTao** [comment]
Update 3

I managed to make a manual update of the plugin. The downloaded and downloadble event are registering without error, but the purchase get stock at "Approved". Also, the AppleID propt for credential pop two time. 

These are the logs I get:

[Log] [store.js] DEBUG: store.queries !! 'ca.tv5.francolabjunior.testWithData registered' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'hosted content download registered' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'non consumable registered' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'registered' (store-ios.js, line 460)
[Log] [store.js] DEBUG: ios -> product ca.tv5.francolabjunior.testWithData registered (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'ca.tv5.francolabjunior.testWithData updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'hosted content download updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'non consumable updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: queries ++ 'hosted content download approved' (store-ios.js, line 460)
[Log] [store.js] DEBUG: queries ++ 'hosted content download downloading' (store-ios.js, line 460)
[Log] [store.js] DEBUG: queries ++ 'hosted content download downloaded' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.trigger -> triggering action refreshed (store-ios.js, line 460)
[Log] [store.js] DEBUG: queries !! 'refreshed' (store-ios.js, line 460)
[Log] [store.js] DEBUG: ios -> initializing storekit (store-ios.js, line 460)
[Log] InAppPurchase[js]: setup ok (store-ios.js, line 794)
[Log] [store.js] INFO: ios -> storekit ready (store-ios.js, line 460)
[Log] [store.js] DEBUG: ios -> loading products (store-ios.js, line 460)
[Log] InAppPurchase[js]: load ["ca.tv5.francolabjunior.testWithData"] (store-ios.js, line 794)
[Log] productState changed to registered (app.js, line 137300)
[Log] Object (app.js, line 137301)

[Log] InAppPurchase[js]: load ok: { valid:[{"id":"ca.tv5.francolabjunior.testWithData","title":"Test avec Data","description":"Test avec Data","price":"$0.00"}] invalid:[] } (store-ios.js, line 794)
[Log] [store.js] DEBUG: ios -> products loaded (store-ios.js, line 460)
[Log] [store.js] DEBUG: ios -> product ca.tv5.francolabjunior.testWithData is valid (hosted content download) (store-ios.js, line 460)
[Log] [store.js] DEBUG: ios -> owned? false (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'ca.tv5.francolabjunior.testWithData valid' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'hosted content download valid' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'non consumable valid' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'valid valid' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'valid' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'ca.tv5.francolabjunior.testWithData updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'hosted content download updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'non consumable updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'valid updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'ca.tv5.francolabjunior.testWithData loaded' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'hosted content download loaded' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'non consumable loaded' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'valid loaded' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'loaded' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'ca.tv5.francolabjunior.testWithData updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'hosted content download updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'non consumable updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'valid updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'updated' (store-ios.js, line 460)
[Log] ===> Store ready!!! (app.js, line 137392)
[Log] [store.js] INFO: ios -> transaction 1000000172014889 purchased (1 in the queue for ca.tv5.francolabjunior.testWithData) (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'ca.tv5.francolabjunior.testWithData approved' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'hosted content download approved' (store-ios.js, line 460)
[Log] You've purchased the content for ca.tv5.francolabjunior.testWithData - it will now download to your device! (app.js, line 137362)
[Log] Object (app.js, line 137363)

[Log] [store.js] DEBUG: store.queries !! 'non consumable approved' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'valid approved' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'approved' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'ca.tv5.francolabjunior.testWithData updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'hosted content download updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'non consumable updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'valid updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'ca.tv5.francolabjunior.testWithData requested' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'hosted content download requested' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'non consumable requested' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'valid requested' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'requested' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'ca.tv5.francolabjunior.testWithData updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'hosted content download updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'non consumable updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'valid updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: ios -> is purchasing ca.tv5.francolabjunior.testWithData (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'ca.tv5.francolabjunior.testWithData initiated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'hosted content download initiated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'non consumable initiated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'valid initiated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'initiated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'ca.tv5.francolabjunior.testWithData updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'hosted content download updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'non consumable updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'valid updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'updated' (store-ios.js, line 460)
[Log] productState changed to initiated (app.js, line 137300)
[Log] Object (app.js, line 137301)

[Log] [store.js] INFO: ios -> transaction 1000000172201618 purchased (2 in the queue for ca.tv5.francolabjunior.testWithData) (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'ca.tv5.francolabjunior.testWithData approved' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'hosted content download approved' (store-ios.js, line 460)
[Log] You've purchased the content for ca.tv5.francolabjunior.testWithData - it will now download to your device! (app.js, line 137362)
[Log] Object (app.js, line 137363)

[Log] [store.js] DEBUG: store.queries !! 'non consumable approved' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'valid approved' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'approved' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'ca.tv5.francolabjunior.testWithData updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'hosted content download updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'non consumable updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'valid updated' (store-ios.js, line 460)
[Log] [store.js] DEBUG: store.queries !! 'updated' (store-ios.js, line 460)
[Log] productState changed to approved (app.js, line 137300)
[Log] Object (app.js, line 137301)

**Comment by G3n3ralTao** [comment]
Update 4

Found the problem and it's working now. While I manually installed the plugin, the iOS NativeCode wasn't copied in the XCode project and was trying to call the old code that didn't have the download procedure.

The main problem is the automatic download via cordova plugin add ... . I updated Cordova, restarted a new project and the file downloaded with the plugin are still not the right one. I tested it on different desktop computer and I still get the same problem. 

Am I the only who got this issue? This is pretty major if people trying to get the plugin get the wrong version.

**Comment by G3n3ralTao** [comment]
Ping?

**Comment by dpa99c** [comment]
Check what version of Cordova you are using: `cordova -v`

Cordova 5.0.0 and above uses npmjs as its plugin registry, but Cordova 3.x/4.x use the Cordova Plugin Registry which is now deprecated and read-only which means plugins on there are no longer maintained. 

Also, if you have an older version of the plugin in your `project/plugins/` folder, then `cordova plugin add` will use the older local copy of the plugin rather than the newer version in the registry. So in order to update a plugin, you must first remove it using `cordova plugin rm` (you can then check it's removed from `plugins/`) before running `cordova plugin add` which will download the latest version from the registry and install it to your project

**Comment by G3n3ralTao** [comment]
I am using the version 5.1.1 of Cordova. I reinstalled Cordova, created a new project with no config file or older version of the plugin in it and "cordova plugin add" still download an incorrect version. Something is wrong somewhere and to be frank it did cost me a lot of time for debugging since when I check the version of the plugin, it was saying 3.11.0.

**Comment by dpa99c** [comment]
Just to be clear:

- exactly what plugin name are you entering at the command line? 
- what is the version number of the incorrect old version you are getting?
- what is the version number you are expecting?


Entering `cordova plugin add cc.fovea.cordova.purchase` will add the "latest" version of the plugin from **this** github repo ([https://github.com/j3k0/cordova-plugin-purchase](https://github.com/j3k0/cordova-plugin-purchase)) via [npmjs](https://www.npmjs.com/package/cc.fovea.cordova.purchase). 

Note that the plugin version is listed on [npmjs](https://www.npmjs.com/package/cc.fovea.cordova.purchase) as 3.11.1 but if you install it into your project and do `cordova plugin ls` you'll see 3.11.0. This is because there's currently a discrepancy between [package.json](https://github.com/j3k0/cordova-plugin-purchase/blob/master/package.json) version of 3.11.1 and the [plugin.xml](https://github.com/j3k0/cordova-plugin-purchase/blob/master/plugin.xml) version of 3.11.0. 
@j3k0 Should we bump both versions to 3.11.2 and `npm publish` to avoid this confusion?

Note that if you enter `cordova plugin add cordova-plugin-purchase`, this will add a plugin from a different github repo - [https://github.com/hypery2k/cordova-purchase-plugin](https://github.com/hypery2k/cordova-purchase-plugin) - the latest version of this one on [npmjs](https://www.npmjs.com/package/cordova-plugin-purchase) is 1.0.4 and it doesn't contain the Apple Hosted Content functionality.

**Comment by G3n3ralTao** [comment]
As I have stated in my previous updated on this topic, I used the command "cordova plugin add cc.fovea.cordova.purchase". I never used "cordova plugin add cordova-plugin-purchase" 

When I check the version number its 3.11.0 but in the code, there are no downloading and downloaded promise on product. My current project depends on the Apple Content for hosting and I was banging my head on the plugin to make it download it after the purchase when there was no downloading procedure in the code.  

Each time I get the latest code from the github and I don't. I can reproduce that problem on a computers with fresh installation of cordova. I had to manually install it on my fresh project to make it work.
